pkg/search/postgres/store_cache.go

```diff
@@ -0,0 +1,386 @@
+		upsertAllowed,
+		targetResource,
+	)
+	store := &cachedStore[T, PT]{
+		schema:          schema,
+		pkGetter:        pkGetter,
+		targetResource:  targetResource,
+		cache:           make(map[string]PT),
+		underlyingStore: underlyingStore,
+
+		setCacheOperationDurationTime: setCacheOperationDurationTime,
+	}
+	// Initial population of the cache. Make sure it is in sync with the DB.
+	store.repopulateCache()
+	return store
+}
+
+// NewGenericStoreWithCacheAndPermissionChecker returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStoreWithCacheAndPermissionChecker[T any, PT clonedUnmarshaler[T]](
+	db postgres.DB,
+	schema *walker.Schema,
+	pkGetter primaryKeyGetter[T, PT],
+	insertInto inserter[T, PT],
+	copyFromObj copier[T, PT],
+	setAcquireDBConnDuration durationTimeSetter,
+	setPostgresOperationDurationTime durationTimeSetter,
+	setCacheOperationDurationTime durationTimeSetter,
+	checker walker.PermissionChecker,
+) Store[T, PT] {
+	underlyingStore := NewGenericStoreWithPermissionChecker[T, PT](
+		db,
+		schema,
+		pkGetter,
+		insertInto,
+		copyFromObj,
+		setAcquireDBConnDuration,
+		setPostgresOperationDurationTime,
+		checker,
+	)
+	store := &cachedStore[T, PT]{
+		schema:            schema,
+		pkGetter:          pkGetter,
+		permissionChecker: checker,
+		cache:             make(map[string]PT),
+		underlyingStore:   underlyingStore,
+
+		setCacheOperationDurationTime: setCacheOperationDurationTime,
+	}
+	// Initial population of the cache. Make sure it is in sync with the DB.
+	store.repopulateCache()
+	return store
+}
+
+// cachedStore implements subset of Store interface for resources with single ID.
+type cachedStore[T any, PT clonedUnmarshaler[T]] struct {
+	schema                        *walker.Schema
+	pkGetter                      primaryKeyGetter[T, PT]
+	setCacheOperationDurationTime durationTimeSetter
+	permissionChecker             walker.PermissionChecker
+	targetResource                permissions.ResourceMetadata
+	underlyingStore               Store[T, PT]
+	cache                         map[string]PT
+	cacheLock                     sync.RWMutex
+}
+
+// Upsert saves the current state of an object in storage.
+func (c *cachedStore[T, PT]) Upsert(ctx context.Context, obj PT) error {
+	dbErr := c.underlyingStore.Upsert(ctx, obj)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Upsert)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	c.addToCacheNoLock(obj.Clone())
+	return nil
+}
+
+// UpsertMany saves the state of multiple objects in the storage.
+func (c *cachedStore[T, PT]) UpsertMany(ctx context.Context, objs []PT) error {
+	dbErr := c.underlyingStore.UpsertMany(ctx, objs)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.UpdateMany)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	for _, obj := range objs {
+		c.addToCacheNoLock(obj.Clone())
```

Comment:
Why do we also need to clone when adding things to the cache? Right now it seems like we clone twice: when "returning" an object to the client, as well as when storing things within the cache (i.e. during upsert).

Seems like we can get away with only cloning once during returning objects.

Reply:
We have to clone twice to be sure to avoid race conditions. The examples are that the caller adds to the cache then changes a value, but another caller is having the cache clone. This would be a write/read conflict and if we don't clone out of the cache, then two callers could have the same reference to the same object